bin/writer.sunxi_raw

```diff
@@ -0,0 +1,36 @@
+# Copyright 2019 Google LLC. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+# Writes sunxi image to the board
+
+# Args:
+#  $1: Mount point of the sdcard when board is off
+#  $2: U-Boot build directory
+
+set -ex
+
+device=$1
+build=$2
+
+echo "Writing to ${device} from build at ${build}"
+
+dd if=/dev/zero of=$device bs=1k count=1024
+dd if=${build}/u-boot-sunxi-with-spl.bin of=$device bs=1024 seek=8
+sync
```

Comment:
dd is synchronous, i.e. does sync internally IIRC.

Reply:
Not in my experience. Do you have a reference?

Also:

https://unix.stackexchange.com/questions/312687/why-is-sync-so-important-when-making-a-bootable-linux-usb-stick

Reply:
I see the synchronous behaviour of dd in practice all the time. I just confirmed it on my Ubuntu 18.04 machine writing a large RPi disk image to a built-in SD card reader on my laptop:

sudo strace dd if=2018-11-13-raspbian-stretch-lite.img of=/dev/mmcblk0 bs=32768 2>&1 | tee log.txt

There's a shortish period where dd alternates between read(0) and write(1) where the data is read from the SSD and pushed towards the SD card via the RAM-based Linux kernel disk cache. When close(1) is called, it synchronously takes a long time (minutes). strace does confirm that dd is not calling fdatasync() or similar; it's literally the close() syscall that's blocking for a long time. I *assume* this is because all the written data is being flushed to the SD card, but haven't looked into the kernel to 100% confirm this. It is plausible there's some kind of non-default sysctl/... that I set on my laptop, or a non-default sysctl/... that Ubuntu sets but other distros don't, that causes this; I don't recall.

Assuming this is standard behaviour (and I haven't confirmed this), then I'd blame that stackexchange post (and many similar articles Google finds) on cargo-cult repetition of a falsehood; that kind of thing does happen pretty frequently. I'll see if I can dig up any real reference.

Reply:
Yes, this is true for block devices at least:

https://elixir.bootlin.com/linux/v5.5/source/fs/block_dev.c#L2145
dev_blk_fops.release = blkdev_close
blkdev_close calls blkdev_put
blkdev_put calls  __blkdev_put
 __blkdev_put calls sync_blockdev if the device open count is 0
sync_blockdev calls __sync_blockdev with the wait flag set
__sync_blockdev calls filemap_write_and_wait

... so that explains the behaviour I'm seeing, at I didn't see any configurability during that code path, so I believe it should apply to all kernels.

Reply:
OK I removed it. That code seems to be the same going back years. Thanks for looking it up.

I often have trouble with a USB device not booting correctly and it seems more reliable with the sync, but perhaps that is just superstition. Let's try it without.